crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -276,20 +270,21 @@ where
                     // Mark the binding in the defining scopes as used too. (Skip the global scope
                     // and the current scope.)
                     for (name, range) in names.iter().zip(ranges.iter()) {
-                        let mut exists = false;
-                        let mut scopes_iter = self.ctx.scope_stack.iter();
-                        // Skip the global scope
-                        scopes_iter.next_back();
-
-                        for index in scopes_iter.skip(1) {
-                            if let Some(index) = self.ctx.scopes[*index].get(name.as_str()) {
-                                exists = true;
-                                self.ctx.bindings[*index].runtime_usage = usage;
-                            }
-                        }
-
-                        // Ensure that every nonlocal has an existing binding from a parent scope.
-                        if !exists {
+                        let binding_id = self
+                            .ctx
+                            .scopes
+                            .ancestors(self.ctx.scope_id)
+                            .skip(1)
+                            .take_while(|scope_id| !scope_id.is_global())
+                            .find_map(|scope_id| {
+                                let scope = &self.ctx.scopes[scope_id];
+                                scope.get(name.as_str())
+                            });
```

Comment:
I would expect that `scopes.ancestors` returns an `Iterator<Item=&Scope>` and not an iterator over the scope ids (`ancestor_ids`?). Returning a `Scope` iterator would hide some of the `id` to scope mapping complexity.